### PR TITLE
fix(tekton): Activity and get build logs should support metapipeline

### LINKED ISSUE #4362: metapipeline: PipelineActivity and jx get build logs need to treat generator and execution PipelineRuns as one pipeline
`PipelineActivity` and `jx get build logs` currently assume every Tekton build has one and only one `PipelineRun` - but that's not going to be the case once the metapipeline (see #4320 and #4361) comes along. We're going to need to modify `PipelineRunInfo` and related code to treat both the CRD generator `PipelineRun` and the resulting actual execution `PipelineRun` as a single build, with proper ordering and results. The user shouldn't need to know that these are technically two separate "builds" behind the scenes - in fact, if the generator `PipelineRun` succeeds, we may not even want to bother exposing it in `PipelineActivity`? In any case, we need to figure this out. =)

cc @hferentschik @joseblas @dgozalo

### PR DESCRIPTION
**Pull request by abayer**
#### Submitter checklist

- [x] Change is code complete and matches issue description.
- [x] Change is covered by existing or new tests.

#### Description

So let's look for all pipelines that match an org/repo/branch/build number, which will give us both the metapipeline and the generated pipeline.

Tests are still needed, just want to see how this, y'know, *builds*.

#### Special notes for the reviewer(s)

/assign @hferentschik 

#### Which issue this PR fixes

fixes #4362

### CODE REVIEW

**Comment by abayer** [comment]
well, that Tekton failure is legit. =)

**Comment by abayer** [comment]
yaaaay Tekton BDD passed. Good sign. =)

**Comment by abayer** [comment]
/hold

Putting this back on hold - it's not working _quite_ right yet.

**Comment by abayer** [comment]
/hold cancel

Ok, I think we're all good.

**Comment by abayer** [comment]
Prow being picky about trailing spaces/newlines after commands is...annoying.

**Comment by hferentschik** [comment]
Nice commit history 👍. I like :-) BTW, you know you don't have to sign-off your commits (adding your email) if you are GPG signing (option `-S`) your commits. Only works for members of jenkins-x. Just only found out recently myself.

File: pkg/cmd/controller/controller_build.go [hunk 1]
```diff
@@ -648,47 +648,53 @@ func (o *ControllerBuildOptions) updatePipelineActivityForRun(kubeClient kuberne
 	if allCompleted {
 		if failed {
 			spec.Status = v1.ActivityStatusTypeFailed
+		} else if pri.Type == tekton.MetaPipeline {
```
Comment (hferentschik) [inline pkg/cmd/controller/controller_build.go]: I am not so familiar with this controller yet, but by looking at the code I am a bit confused by seeing something like `if allCompleted` and then  `ActivityStatusTypePending` if the type is tekton.MetaPipeline. Not quite sure how to do it differently atm though.

Comment (hferentschik) [reply pkg/cmd/controller/controller_build.go]: would it make more sense to do the switch between the type once at the beginning of the function and handle the case of meta pipeline or execution pipeline?

File: pkg/cmd/get/get_build_logs.go [hunk 2]
```diff
@@ -233,7 +233,8 @@ func (o *GetBuildLogsOptions) getLastJenkinsBuild(name string, buildNumber int)
 	}
 }
 
-func (o *GetBuildLogsOptions) getProwBuildLog(kubeClient kubernetes.Interface, tektonClient tektonclient.Interface, jxClient versioned.Interface, ns string, tektonEnabled bool) error {
+// GetProwBuildLog prompts the user, if needed, to choose a pipeline, and then prints out that pipeline's logs.
+func (o *GetBuildLogsOptions) GetProwBuildLog(kubeClient kubernetes.Interface, tektonClient tektonclient.Interface, jxClient versioned.Interface, ns string, tektonEnabled bool) error {
```
Comment (hferentschik) [inline pkg/cmd/get/get_build_logs.go]: So how does this change pan out in the end? `jx get build log` lists meta pipeline build as different items for which the logs can be retrieved?

Comment (hferentschik) [reply pkg/cmd/get/get_build_logs.go]: How can I see how this looks like in the end?

File: pkg/cmd/get/get_build_logs.go [hunk 3]
```diff
@@ -322,90 +322,98 @@ func (o *GetBuildLogsOptions) getProwBuildLog(kubeClient kubernetes.Interface, t
...
-			}
-			if stage.Pod == nil {
-				// The stage's pod hasn't been created yet, so let's wait a bit.
-				f := func() error {
-					selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{
-						pipeline.GroupName + pipeline.PipelineRunLabelKey: pri.PipelineRun,
-					}})
-					if err != nil {
-						return err
+		var pipelines []*tekton.PipelineRunInfo
```
Comment (hferentschik) [inline pkg/cmd/get/get_build_logs.go]: If it were me I would for sure extract some functions here. This is ways too nested for my taste.

File: pkg/cmd/get/get_build_logs.go [hunk 4]
```diff
@@ -505,27 +521,26 @@ func (o *GetBuildLogsOptions) loadBuilds(kubeClient kubernetes.Interface, ns str
...
 			defaultName = name
 		}
 	}
-	return names, defaultName, buildMap, pipelineMap, nil
+	return names, defaultName, buildMap, nil
 }
 
-func (o *GetBuildLogsOptions) loadPipelines(kubeClient kubernetes.Interface, tektonClient tektonclient.Interface, jxClient versioned.Interface, ns string) ([]string, string, map[string]builds.BaseBuildInfo, map[string]builds.BaseBuildInfo, error) {
+// LoadPipelines loads all available pipelines as PipelineRunInfos.
+func (o *GetBuildLogsOptions) LoadPipelines(kubeClient kubernetes.Interface, tektonClient tektonclient.Interface, jxClient versioned.Interface, ns string) ([]string, string, map[string][]builds.BaseBuildInfo, error) {
```
Comment (hferentschik) [inline pkg/cmd/get/get_build_logs.go]: Why do you need to export this function? Is it just for testing? If so, then I think this is an anti-pattern. Then I rather put the test into the same package.

File: pkg/cmd/get/get_build_logs_test.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package get_test
```
Comment (hferentschik) [inline pkg/cmd/get/get_build_logs_test.go]: I'd use package `get`. I think it is fully legit to test none exported functions.

File: pkg/cmd/controller/controller_build.go (hunk 1 above)
Comment (abayer) [reply pkg/cmd/controller/controller_build.go]: `allCompleted` refers to the current executing `PipelineRun`, not the `PipelineActivity` as a whole. We want to treat the meta and execution `PipelineRun`s both as part of the same `PipelineActivity`, so if the meta pipeline didn't fail, we're expecting there to be an execution pipeline to come that we'll add to the `PipelineActivity`.

File: pkg/cmd/get/get_build_logs_test.go (hunk 5 above)
Comment (abayer) [reply pkg/cmd/get/get_build_logs_test.go]: Will do!

File: pkg/cmd/get/get_build_logs.go (hunk 2 above)
Comment (abayer) [reply pkg/cmd/get/get_build_logs.go]: `jx get build log` will treat the meta pipeline and execution pipeline as a single build, with the logs for all steps in both of them included in the proper order.

File: pkg/cmd/controller/controller_build.go (hunk 1 above)
Comment (hferentschik) [reply pkg/cmd/controller/controller_build.go]: Sure. So it is more a "allStageCompleted"

Comment (abayer) [reply pkg/cmd/controller/controller_build.go]: Yeah, `allStagesCompleted` feels right. I'll rename it.

File: pkg/cmd/controller/controller_build.go [hunk 6]
```diff
@@ -606,7 +606,7 @@ func (o *ControllerBuildOptions) updatePipelineActivityForRun(kubeClient kuberne
 	spec := &activity.Spec
 	var biggestFinishedAt metav1.Time
 
-	allCompleted := true
+	allStagesCompleted := true
```
Comment (hferentschik) [inline pkg/cmd/controller/controller_build.go]: :-)

File: pkg/cmd/get/get_build_logs_test.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
+package get
```
Comment (hferentschik) [inline pkg/cmd/get/get_build_logs_test.go]: 👍

**Pull request merged**